### ISSUE TITLE
pg.image.load(fileobj) doesn't work in pygame 2

### ISSUE THREAD

**Issue by DeluxeYang** [pygame/pygame#1516]
```
jpg_file = BytesIO(jpg_bytes)
unit_image = pg.image.load(jpg_file).convert()
```

this code works in pygame 1.9.6, but return `pygame.error: Unsupported image format` in pygame 2 dev5

**Comment by robertpfeiffer** [comment]
I could not reproduce your problem. Could you post the JPEG file?

I tried this:
```
import pygame
import io
import sys
pygame.init()
screen=pygame.display.set_mode((640,480))

filename=sys.argv[1]

contents = open(filename, "rb").read()
f_bytes = io.BytesIO(contents)

f_opened = open(filename, "rb")

img0=pygame.image.load(filename).convert()
img1=pygame.image.load(f_opened).convert()
img2=pygame.image.load(f_bytes).convert()

screen.blit(img0, (0,0))
pygame.display.flip()
pygame.time.delay(1000)

screen.blit(img1, (0,0))
pygame.display.flip()
pygame.time.delay(1000)

screen.blit(img2, (0,0))
pygame.display.flip()
pygame.time.delay(1000)
```

**Comment by Starbuck5** [comment]
I could reproduce on Win10.

It seems to be just jpgs, I tested gif, tif, and png without issue.

Output:
```
pygame 2.0.0.dev6 (SDL 2.0.10, python 3.7.5)
Hello from the pygame community. https://www.pygame.org/contribute.html
Traceback (most recent call last):
  File "C:\Users\Starbuck\Desktop\test.py", line 16, in <module>
    img2=pygame.image.load(f_bytes).convert()
pygame.error: Unsupported image format
```

**Comment by slackerbob** [comment]
I can also confirm this issue in dev10

```
res = urlopen(imgurl, timeout=10)
image_str = res.read()
res.close()
image_file = io.BytesIO(image_str)
img_logo = pygame.image.load(image_file)

```
results in:

`pygame.error: Unsupported image format`

I was banging my head against the wall until I realized it worked fine in pygame 1.9.6

**Comment by illume** [comment]
Is this with all image urls or one in particular?

**Comment by slackerbob** [comment]
For me it's happening on all .jpg BytesIO objects and not on any .png BytesIO objects.

BUT I've been testing this and so far I can only reproduce this issue on **Raspbian Buster**.

I could **not** reproduce the issue on OSX Catalina, nor on Arch Linux (x64).

I'd be curious to know if the OP had this issue on Raspbian specifically, also.

**Comment by robertpfeiffer** [comment]
@slackerbob What SDL2 version are you using on raspbian?

**Comment by robertpfeiffer** [comment]
To those who suffer from this bug, can you please try using the file name parameter and try again?
The docs even say: `load(fileobj, namehint="")`, and I'd be inclined to print a warning going forward when no dummy filename is provided via namehint to infer the data type. Maybe we could have a mime type parameter also.

> Pygame will automatically determine the image type (e.g., GIF or bitmap) and create a new Surface object from the data. In some cases it will need to know the file extension (e.g., GIF images should end in ".gif"). If you pass a raw file-like object, you may also want to pass the original filename as the namehint argument.

It is very unfortunate that this was allowed in the first place, now we need to live with the backward compatibility.

**Comment by slackerbob** [comment]
Thank you for the help!

SDL version: 2.0.9
Python Version: 2.7.3

New code:
```
res = urlopen(imgurl, timeout=10)
image_str = res.read()
res.close()
image_file = io.BytesIO(image_str)

filepath = urlparse(imgurl)
filename = os.path.basename(filepath.path)

img_logo = pygame.image.load(image_file, namehint=filename)
```

And for every image type (I've tested png and jpg) pygame.image.load throws the error:
`load_extended() takes no keyword arguments
`

**Comment by slackerbob** [comment]
After sorting through the pygame source a bit, I noticed that load_extended seems to be a pointer to:
```
static PyObject *
image_load_ext(PyObject *self, PyObject *arg)
```

Which doesn't have named arguments, so I dropped the "namehint=" part of the command, so it's just:
`img_logo = pygame.image.load(image_file, filename)
`

Which unfortunately, results in the original exception:
`pygame.error: Unsupported image format
`
for jpg files (but once again loads png files fine).

**Comment by MyreMylar** [comment]
> SDL version: 2.0.9
Python Version: 2.7.3

These are both fairly old. I wonder if this issue vanishes if you use python 3, or if you upgrade SDL to 2.0.12

**Comment by MyreMylar** [comment]
Raspbian/Raspberry Pi OS (the name changed last year btw so you may want to update your OS too) should already have python 3 installed.

**Comment by MyreMylar** [comment]
Though we also had one report on windows 10 too...

**Comment by MyreMylar** [comment]
Just tried it and was able to load and display a jpeg on windows 10, so not working on windows appears not to be universal.

This was the code I used:

```
import pygame

pygame.display.init()

screen = pygame.display.set_mode((640, 480))

image = pygame.image.load('blue_rect.jpg')

screen.blit(image, (0, 0))
pygame.display.update()
running = True

while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
```

And this jpg:
![blue_rect](https://user-images.githubusercontent.com/13382426/87159978-162d6200-c2ba-11ea-96ea-71e66aabb0b8.jpg)

**Comment by MyreMylar** [comment]
Oh hang on, this is only when loading from file objects? Lemme re-check.

**Comment by MyreMylar** [comment]
OK, yes I can confirm that this code:

```
import io

import pygame

pygame.display.init()

screen = pygame.display.set_mode((640, 480))

jpg_bytes = open('blue_rect.jpg', "rb").read()
jpg_file = io.BytesIO(jpg_bytes)

image = pygame.image.load(jpg_file)

screen.blit(image, (0, 0))
pygame.display.update()
running = True

while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
```

Also errors on windows 10 with pygame 2.0.0.dev10 (SDL 2.0.12, python 3.8.2). adding name hints (jpg, .jpg, .jpeg) seems to have no noticeable effect.

**Comment by MyreMylar** [comment]
So, looking through the code the error bubbles up from the C code in SDL_image, I'm guessing that the fault lies in the `IMG_isJPG` function from here:

https://hg.libsdl.org/SDL_image/file/59a4c1ea96f3/IMG_jpg.c

Which includes this warning:

> /* This detection code is by Steaphan Greene <stea@cs.binghamton.edu> */
> /* Blame me, not Sam, if this doesn't work right. */
> /* And don't forget to report the problem to the the sdl list too! */

As far as I can tell that detection code is always run by this part of the SDL_image code:

```
    /* Detect the type of image being loaded */
    for ( i=0; i < SDL_arraysize(supported); ++i ) {
        if (supported[i].is) {
            if (!supported[i].is(src))
                continue;
        } else {
            /* magicless format */
            if (!type || !IMG_string_equals(type, supported[i].type))
                continue;
        }
#ifdef DEBUG_IMGLIB
        fprintf(stderr, "IMGLIB: Loading image as %s\n",
            supported[i].type);
#endif
        image = supported[i].load(src);
        if (freesrc)
            SDL_RWclose(src);
        return image;
    }

    if ( freesrc ) {
        SDL_RWclose(src);
    }
    IMG_SetError("Unsupported image format");
    return NULL;
}
```
It looks like it goes through all the image type detection functions one after the other whether you provide a type hint or not, so that means the jpg one must be returning False in the case of being passed binary data.

**Comment by slackerbob** [comment]
> > SDL version: 2.0.9
> > Python Version: 2.7.3
> 
> These are both fairly old. I wonder if this issue vanishes if you use python 3, or if you upgrade SDL to 2.0.12

I apologize. That was a typo. It's running python version **3**.7.3

**Comment by MyreMylar** [comment]
I think the best path forward would be to to try and prove that this issue exists in the SDL C code on windows or raspberry Pi and then report it onwards to the SDL bugzilla. 

The function to check would be `IMG_LoadTyped_RW` I'm not quite enough of a C ninja anymore to know immediately exactly how to get the data into the same format as what python is passing through. It looks like the pygame code is just taking the BytesIO object and doing some casting with this macro here:

https://github.com/pygame/pygame/blob/cbfe56227a2eb777f236dced848762bdb5ca1467/src_c/include/_pygame.h#L518-L520

before passing it into `IMG_LoadTyped_RW`

So perhaps this page might provide C equivalents?

https://wiki.libsdl.org/SDL_RWops

**Comment by illume** [comment]
I added the critical label, since this is a regression from pygame 1.9.6

**Comment by ankith26** [comment]
Hmm I’m not too sure, but I feel that my PR #2226 might accidentally fix this issue

**Comment by illume** [comment]
32bit builds are failing on appveyor.

I'm now reading through rwobject.c... there's quite a few places things could go wrong which don't include SDL. I'll confirm everything is going through to SDL or not, and post results in the issue.

**Comment by ankith26** [comment]
It’s interesting, how windows and raspberry pi have nothing much in common, but both are 32-bit OS and that’s why the error is mostly happening

**Comment by ankith26** [comment]
Was just playing around to debug this.
First, I wanted to figure out whether this is an SDL bug or a pygame bug.

So I took out my pi 3b, and put this code in ``imageext.c``
```
static PyObject *
_test_is_jpg(PyObject *self, PyObject *obj)
{
    SDL_RWops *rw = pgRWops_FromFileObject(obj);
    if (!rw)
        return NULL;
    return PyBool_FromLong(IMG_isJPG(rw));
}
```
The goal is to figure out whether ``IMG_isJPG`` is broken or not.
And I wrote this test code
```python
import io

import pygame
import pygame.imageext as ext

with open("delete.jpg", "rb") as f:
    print(ext._test_is_jpg(f))
    print(ext._test_is_jpg(io.BytesIO(f.read())))
```

Its output was very interesting
```
pygame 2.0.2.dev1 (SDL 2.0.9, Python 3.7.3)
Hello from the pygame community. https://www.pygame.org/contribute.html
True
False
```
As you can clearly see, it recognises the python file-like object as a JPG, but fails to recognise the BytesIO object. So the issue is on our side, our rwobject code is not handling stuff well on 32 bit systems

**Comment by ankith26** [comment]
After breaking my head for many hours. I HAVE FINALLY FIXED THIS. It had almost driven me bananas, but I think I am OK now.
The solution was a simple cast of a variable to an int.
Apart from that, I have done a lot of cleanups and refactor. PR coming soon :)

**Comment by slackerbob** [comment]
> After breaking my head for many hours. I HAVE FINALLY FIXED THIS. It had almost driven me bananas, but I think I am OK now.
> The solution was a simple cast of a variable to an int.
> Apart from that, I have done a lot of cleanups and refactor. PR coming soon :)

After more than a month of running your branch with these fixes, I can confirm that this completely solves the issue on a Raspberry Pi running Raspbian Buster.

Once that pull request is finished up and merged, this issue will be solved!

Until that time, the ankith26-misc-fixes branch works for my needs.

Thank you for all your effort on this!

**Comment by ankith26** [comment]
Good to hear that the branch is working well for you. :)

**Comment by TechnoX** [comment]
I'm facing the same issue. 
Is this finally merged into the official pygame repo? Is it released in any recent version?

**Comment by ankith26** [comment]
It is not merged yet, but hopefully it will be merged soon, and maybe it could make it by next pygame version
